### PR TITLE
 [change] e2e test for local jdk and remote jdk

### PR DESCRIPTION
**Pull request by basakkapusuzoglu**
BazelBspJava8Test and BazelBspJava11Test are converted into LocalJdkTest and RemoteJdkTest

---

```
bazel run //e2e:BazelBspLocalJdkTest  
bazel run //e2e:BazelBspRemoteJdkTest
```

### CODE REVIEW

File: e2e/BUILD [hunk 1]
```diff
@@ -38,11 +38,11 @@ sh_binary(
 )
 
 sh_binary(
-    name = "BazelBspJava11ProjectTest",
+    name = "RemoteJdkTest",
```
Comment (abrams27) [inline e2e/BUILD:41]: lets keep the convention of `BazelBsp*Test`

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BUILD [hunk 2]
```diff
@@ -51,8 +51,8 @@ java_binary(
 )
 
 java_binary(
-    name = "BazelBspJava11ProjectTest",
-    srcs = ["BazelBspJava11ProjectTest.java"],
+    name = "RemoteJdkTest",
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BUILD:54]: same

Comment (abrams27) [reply e2e/src/main/java/org/jetbrains/bsp/bazel/BUILD:54]: + there are additional places where the old label is used

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -90,7 +90,7 @@ jobs:
       - name: Set up JDK 11
         uses: actions/setup-java@v1
         with:
-          java-version: 11
+          java-version: 8
```
Comment (abrams27) [inline .github/workflows/integration-tests.yml]: it should be done in remote jdk test

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-tests.yml [hunk 4]
```diff
@@ -82,44 +82,44 @@ jobs:
...
+
+  remote-jdk-test:
+    runs-on: ubuntu-latest
+    needs: setup
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
+        uses: actions/setup-java@v1
+        with:
+          java-version: 11
```
Comment (abrams27) [inline .github/workflows/integration-tests.yml:112]:
```suggestion
          java-version: 8
```

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -82,44 +82,44 @@ jobs:
...
-#          key: ${{ runner.os }}-${{ env.cache-name }}-${{ github.sha }}
-#      - name: Run test
-#        run: bazel run //e2e:BazelBspJava11ProjectTest
-#
+  local-jdk-test:
+    runs-on: ubuntu-latest
+    needs: setup
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
```
Comment (abrams27) [inline .github/workflows/integration-tests.yml:90]:
```suggestion
      - name: Set up JDK 17
```

File: .github/workflows/integration-tests.yml [hunk 6]
```diff
@@ -82,44 +82,44 @@ jobs:
...
+          key: ${{ runner.os }}-${{ env.cache-name }}-${{ github.sha }}
+      - name: Run test
+        run: bazel run //e2e:LocalJdkTest
+
+  remote-jdk-test:
+    runs-on: ubuntu-latest
+    needs: setup
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
```
Comment (abrams27) [inline .github/workflows/integration-tests.yml:109]:
```suggestion
      - name: Set up JDK 8
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/LocalJdkTest.java [hunk 7]
```diff
@@ -11,20 +11,19 @@
 import java.util.List;
 import org.jetbrains.bsp.bazel.base.BazelBspTestBaseScenario;
 import org.jetbrains.bsp.bazel.base.BazelBspTestScenarioStep;
-import org.jetbrains.bsp.bazel.commons.Constants;
 
-public class BazelBspJava8ProjectTest extends BazelBspTestBaseScenario {
+public class LocalJdkTest extends BazelBspTestBaseScenario {
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/LocalJdkTest.java:15]:
```suggestion
public class BazelBspLocalJdkTest extends BazelBspTestBaseScenario {
```

File: server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt [hunk 8]
```diff
@@ -223,7 +223,7 @@ class DiagnosticsServiceTest {
...
...cs //executioncontext/projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/model:model //install/src/main/java/org/jetbrains/bsp/bazel/install/cli:cli //executioncontext/workspacecontext/src/test/java/org/jetbrains/bsp/bazel/workspacecontext/entries/mappers:WorkspaceContextTargetsEntityMapperTest //server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/utils:utils //executioncontext/installationcontext/src/main/java/org/jetbrains/bsp/bazel/installationcontext:installationcontext //server/src/test/java/org/jetbrains/bsp/bazel/server/bsp/utils:SourceRootGuesserTest //install:install //server/src/test/java/org/jetbrains/bsp/bazel/server/sync/dependencytree:DependencyTreeTest //server/src/main/java/org/jetbrains/bsp/bazel/server/sync/proto:bsp_target_info_proto //server/src/main/java/org/jetbrains/bsp/bazel/server/sync:sync //executioncontext/projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/parser:parser //server/src/main/java/org/jetbrains/bsp/bazel:bsp-lib //bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/params:params //commons/src/test/java/org/jetbrains/bsp/bazel/commons:FormatTest //e2e:all //e2e/src/main/java/org/jetbrains/bsp/bazel/base:base
```
Comment (abrams27) [inline server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt:226]: please undo the formatting

**Review by abrams27 — CHANGES_REQUESTED** [review]

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-tests.yml (hunk 4 above)
Comment (basakkapusuzoglu) [reply .github/workflows/integration-tests.yml:112]: it is done I guess it is not updated in git

File: server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt (hunk 8 above)
Comment (basakkapusuzoglu) [reply server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt:226]: I didn't do formatting here

File: .github/workflows/integration-tests.yml (hunk 5 above)
Comment (abrams27) [reply .github/workflows/integration-tests.yml:90]: bump

File: .github/workflows/integration-tests.yml (hunk 4 above)
Comment (abrams27) [reply .github/workflows/integration-tests.yml:112]: are u sure?

File: server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt (hunk 8 above)
Comment (abrams27) [reply server/src/test/java/org/jetbrains/bsp/bazel/server/diagnostics/DiagnosticsServiceTest.kt:226]: did u change something here? if not just revert the changes

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        ) { testClient.testWorkspaceTargets(Duration.ofSeconds(60), workspaceBuildTargetsResult) }
+    }
+
+    companion object {
+        private const val REPO_NAME = "remote-jdk-project"
+
+        // we cannot use `bazel test ...` because test runner blocks bazel daemon,
+        // but testing server needs it for queries and etc
+        @JvmStatic
+        fun main(args: Array<String>) {
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:46]: it doesnt have to be inside companion object

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        ) { testClient.testWorkspaceTargets(Duration.ofSeconds(30), workspaceBuildTargetsResult) }
+    }
+
+    companion object {
+        private const val REPO_NAME = "local-jdk-project"
+
+        // we cannot use `bazel test ...` because test runner blocks bazel daemon,
+        // but testing server needs it for queries and etc
+        @JvmStatic
+        fun main(args: Array<String>) {
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:53]: same

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = ImmutableList.of(workspaceBuildTargets())
+
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+        // TODO this was always broken. We resolve local_jdk which is actual jdk installed
+        // on the running machine. In my case as well as on CI it is Java 11. We were
+        // returning this path annotated as Java 8, but actually it was java 11 anyway.
+        // now it is detected because the version is actually checked rather than inferred
+        // from heuristics.
+        // We should figure out a way to enforce bazel to download runtime jdk that matches
+        // expected number.
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:20]: this comment is no longer actual

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+import ch.epfl.scala.bsp4j.*
+import com.google.common.collect.ImmutableList
+import org.jetbrains.bsp.bazel.base.BazelBspTestBaseScenario
+import org.jetbrains.bsp.bazel.base.BazelBspTestScenarioStep
+import java.time.Duration
+
+class BazelBspRemoteJdkTest : BazelBspTestBaseScenario(REPO_NAME) {
+
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = ImmutableList.of(workspaceBuildTargets())
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:11]:
```suggestion
    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = listOf(workspaceBuildTargets())
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+        val exampleExampleJvmBuildTarget = JvmBuildTarget("file://\$BAZEL_CACHE/external/remotejdk11_linux/", "11")
+        val rootBuildTarget = BuildTarget(
+                BuildTargetIdentifier("bsp-workspace-root"),
+                ImmutableList.of(),
+                ImmutableList.of(),
+                ImmutableList.of(),
+                BuildTargetCapabilities(false, false, false, false))
+        rootBuildTarget.displayName = "bsp-workspace-root"
+        rootBuildTarget.baseDirectory = "file://\$WORKSPACE/"
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:22]: new line

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+class BazelBspLocalJdkTest : BazelBspTestBaseScenario(REPO_NAME) {
+
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = ImmutableList.of(workspaceBuildTargets())
+
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+
+        val exampleExampleJvmBuildTarget = JvmBuildTarget("file://\$BAZEL_CACHE/external/local_jdk/", "17")
+        val rootBuildTarget = BuildTarget(
+                BuildTargetIdentifier("bsp-workspace-root"),
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:18]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:19]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:20]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+                BuildTargetCapabilities(false, false, false, false))
+        rootBuildTarget.displayName = "bsp-workspace-root"
+        rootBuildTarget.baseDirectory = "file://\$WORKSPACE/"
+        val exampleExampleBuildTarget = BuildTarget(
+                BuildTargetIdentifier("//example:example"),
+                ImmutableList.of("application"),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:26]:
```suggestion
                listOf("application"),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+                ImmutableList.of("java"),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:27]:
```suggestion
                listOf("java"),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:28]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+class BazelBspRemoteJdkTest : BazelBspTestBaseScenario(REPO_NAME) {
+
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = ImmutableList.of(workspaceBuildTargets())
+
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+        val exampleExampleJvmBuildTarget = JvmBuildTarget("file://\$BAZEL_CACHE/external/remotejdk11_linux/", "11")
+        val rootBuildTarget = BuildTarget(
+                BuildTargetIdentifier("bsp-workspace-root"),
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:17]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:18]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:19]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+                BuildTargetCapabilities(false, false, false, false))
+        rootBuildTarget.displayName = "bsp-workspace-root"
+        rootBuildTarget.baseDirectory = "file://\$WORKSPACE/"
+        val exampleExampleBuildTarget = BuildTarget(
+                BuildTargetIdentifier("//example:example"),
+                ImmutableList.of("application"),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:25]:
```suggestion
                listOf("application"),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+                ImmutableList.of("java"),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:26]:
```suggestion
                listOf("java"),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+                ImmutableList.of(),
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:27]:
```suggestion
                emptyList(),
```

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+
+        // we cannot use `bazel test ...` because test runner blocks bazel daemon,
+        // but testing server needs it for queries and etc
+        @JvmStatic
+        fun main(args: Array<String>) {
+            val test = BazelBspRemoteJdkTest()
+            test.executeScenario()
+        }
+    }
+}
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:51]: new line

**Review by abrams27 — CHANGES_REQUESTED** [review]

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 27]
```diff
@@ -0,0 +1,52 @@
+package org.jetbrains.bsp.bazel
+
+import ch.epfl.scala.bsp4j.*
+import org.jetbrains.bsp.bazel.base.BazelBspTestBaseScenario
+import org.jetbrains.bsp.bazel.base.BazelBspTestScenarioStep
+import java.time.Duration
+
+class BazelBspLocalJdkTest : BazelBspTestBaseScenario(REPO_NAME) {
+
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = listOf(workspaceBuildTargets())
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:10]: add empty line after the method

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+
+    private fun workspaceBuildTargets(): BazelBspTestScenarioStep {
+
```
[the remark this answers is not in the document]
Comment (abrams27) [reply e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:12]: remove the empty line at the beginning of the method

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+        exampleExampleBuildTarget.data = exampleExampleJvmBuildTarget
+        exampleExampleBuildTarget.dataKind = BuildTargetDataKind.JVM
+        val workspaceBuildTargetsResult = WorkspaceBuildTargetsResult(
+                listOf(rootBuildTarget, exampleExampleBuildTarget))
+        return BazelBspTestScenarioStep(
+                "local-jdk-project workspace build targets"
+        ) { testClient.testWorkspaceTargets(Duration.ofSeconds(30), workspaceBuildTargetsResult) }
+    }
+
+
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspLocalJdkTest.kt:40]: remove one empty line

File: e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt [hunk 30]
```diff
@@ -0,0 +1,50 @@
+package org.jetbrains.bsp.bazel
+
+import ch.epfl.scala.bsp4j.*
+import org.jetbrains.bsp.bazel.base.BazelBspTestBaseScenario
+import org.jetbrains.bsp.bazel.base.BazelBspTestScenarioStep
+import java.time.Duration
+
+class BazelBspRemoteJdkTest : BazelBspTestBaseScenario(REPO_NAME) {
+
+    override fun getScenarioSteps(): List<BazelBspTestScenarioStep> = listOf(workspaceBuildTargets())
```
Comment (abrams27) [inline e2e/src/main/java/org/jetbrains/bsp/bazel/BazelBspRemoteJdkTest.kt:10]: add empty line after the method

**Review by abrams27 — CHANGES_REQUESTED** [review]

**Review by abrams27 — APPROVED** [review]

**Pull request merged**